File: apps/trader-service/app/services/strategy_service.py
from __future__ import annotations

from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import Strategy


class StrategyService:
    async def list_for_user(self, session: AsyncSession, *, user_id: str) -> list[Strategy]:
        result = await session.execute(
            select(Strategy).where(Strategy.user_id == user_id).order_by(Strategy.created_at.desc())
        )
        return list(result.scalars().all())

    async def list_available(self, session: AsyncSession, *, user_id: str) -> list[Strategy]:
        result = await session.execute(
            select(Strategy)
            .where(or_(Strategy.is_public.is_(True), Strategy.user_id == user_id))
            .order_by(Strategy.created_at.desc())
        )
        return list(result.scalars().all())

    async def create(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        name: str,
        description: str,
        tag: str,
        code: str,
        params: dict,
        version: str,
        status: str,
        is_public: bool,
    ) -> Strategy:
        strategy = Strategy(
            user_id=user_id,
            name=name,
            description=description,
            tag=tag,
            code=code,
            params=params,
            version=version,
            status=status,
            is_public=is_public,
        )
        session.add(strategy)
        await session.commit()
        await session.refresh(strategy)
        return strategy

    async def get(self, session: AsyncSession, *, user_id: str, strategy_id: str) -> Strategy:
        strategy = await session.get(Strategy, strategy_id)
        if not strategy:
            raise ValueError("Strategy not found")
        if strategy.user_id != user_id and not strategy.is_public:
            raise ValueError("Strategy not found")
        return strategy

    async def update(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        strategy_id: str,
        **changes,
    ) -> Strategy:
        strategy = await session.get(Strategy, strategy_id)
        if not strategy or strategy.user_id != user_id:
            raise ValueError("Strategy not found")

        for key, value in changes.items():
            if value is None or not hasattr(strategy, key):
                continue
            setattr(strategy, key, value)

        await session.commit()
        await session.refresh(strategy)
        return strategy

    async def delete(self, session: AsyncSession, *, user_id: str, strategy_id: str) -> None:
        strategy = await session.get(Strategy, strategy_id)
        if not strategy or strategy.user_id != user_id:
            raise ValueError("Strategy not found")
        await session.delete(strategy)
        await session.commit()

File: apps/trader-service-node/src/service/strategy.service.ts
import { Provide, httpError } from '@midwayjs/core';
import { InjectEntityModel } from '@midwayjs/typeorm';
import type { Repository } from 'typeorm';
import { Strategy } from '../entity/strategy.entity.js';

type StrategyCreateParams = {
  userid: number;
  name: string;
  description?: string;
  tag?: string;
  code?: string;
  params?: Record<string, unknown>;
  version?: string;
  status?: string;
  isPublic: boolean;
};

type StrategyUpdateParams = {
  name?: string | null;
  description?: string | null;
  tag?: string | null;
  code?: string | null;
  params?: Record<string, unknown> | null;
  version?: string | null;
  status?: string | null;
  isPublic?: boolean | null;
};

@Provide()
export class StrategyService {
  @InjectEntityModel(Strategy)
  strategyRepo?: Repository<Strategy>;

  private requireRepo(): Repository<Strategy> {
    if (!this.strategyRepo) {
      throw new httpError.ServiceUnavailableError('Database not configured');
    }
    return this.strategyRepo;
  }

  async listForUser(userid: number): Promise<Strategy[]> {
    const repo = this.requireRepo();
    return repo.find({ 
      where: { userid }, 
      order: { createdAt: 'DESC' },
      relations: ['user']
    });
  }

  async listAvailable(userid: number): Promise<Strategy[]> {
    const repo = this.requireRepo();
    return repo
      .createQueryBuilder('strategy')
      .leftJoinAndSelect('strategy.user', 'user')
      .where('strategy.isPublic = :isPublic', { isPublic: true })
      .orWhere('strategy.userid = :userid', { userid })
      .orderBy('strategy.createdAt', 'DESC')
      .getMany();
  }

  async create(params: StrategyCreateParams): Promise<Strategy> {
    const repo = this.requireRepo();
    const strategy = repo.create({
      userid: params.userid,
      name: params.name,
      description: params.description ?? '',
      tag: params.tag ?? 'neutral',
      code: params.code ?? '',
      params: params.params ?? {},
      version: params.version ?? 'v1',
      status: params.status ?? 'inactive',
      isPublic: params.isPublic,
    });
    return await repo.save(strategy);
  }

  async get(userid: number, strategyId: number): Promise<Strategy> {
    const repo = this.requireRepo();
    const strategy = await repo.findOne({ 
      where: { id: strategyId },
      relations: ['user']
    });
    if (!strategy) {
      throw new httpError.NotFoundError('Strategy not found');
    }
    if (strategy.userid !== userid && !strategy.isPublic) {
      throw new httpError.NotFoundError('Strategy not found');
    }
    return strategy;
  }

  async update(userid: number, strategyId: number, patch: StrategyUpdateParams): Promise<Strategy> {
    const repo = this.requireRepo();
    const strategy = await repo.findOne({ where: { id: strategyId } });
    if (!strategy || strategy.userid !== userid) {
      throw new httpError.NotFoundError('Strategy not found');
    }

    if (patch.name !== undefined && patch.name !== null) strategy.name = patch.name;
    if (patch.description !== undefined && patch.description !== null) strategy.description = patch.description;
    if (patch.tag !== undefined && patch.tag !== null) strategy.tag = patch.tag;
    if (patch.code !== undefined && patch.code !== null) strategy.code = patch.code;
    if (patch.params !== undefined && patch.params !== null) strategy.params = patch.params;
    if (patch.version !== undefined && patch.version !== null) strategy.version = patch.version;
    if (patch.status !== undefined && patch.status !== null) strategy.status = patch.status;
    if (patch.isPublic !== undefined && patch.isPublic !== null) strategy.isPublic = patch.isPublic;

    return await repo.save(strategy);
  }

  async delete(userid: number, strategyId: number): Promise<void> {
    const repo = this.requireRepo();
    const strategy = await repo.findOne({ where: { id: strategyId } });
    if (!strategy || strategy.userid !== userid) {
      throw new httpError.NotFoundError('Strategy not found');
    }
    await repo.remove(strategy);
  }
}
